File: backend/queryview/connect.py
"""Connection domain: the SQLModel/SQLite connection store (passwords encrypted
at rest) and per-session active connections. No HTTP-server concerns live here;
the operations return plain results and main.py maps them to responses."""

from __future__ import annotations

import base64
import os
import time
from collections import OrderedDict
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from sqlalchemy.ext.asyncio import create_async_engine
from sqlmodel import Field, SQLModel, select
from sqlmodel.ext.asyncio.session import AsyncSession

from .clickhouse import (
    ChConfig,
    ch_query,
    describe_query as ch_describe_query,
    list_databases,
)

# --- Storage (SQLite, lazily opened) --------------------------------------


class Connection(SQLModel, table=True):
    __tablename__ = "connections"

    id: int | None = Field(default=None, primary_key=True)
    name: str = Field(unique=True, index=True)
    type: str = Field(default="clickhouse", index=True)
    host: str
    port: int
    username: str
    password: str  # encrypted at rest, never plaintext
    database: str | None = Field(default=None)
    last_active_at: int  # unix ms; the max is the "latest active"


def _db_path() -> Path:
    env = os.environ.get("DB_PATH")
    if env:
        return Path(env)
    return Path(__file__).resolve().parent.parent / "queryview.db"


_engine = None
_schema_ready = False


def _engine_for_db():
    """The async SQLAlchemy engine (aiosqlite), memoized. Lazy so importing this
    module has no side effects — no file is touched until the first query."""
    global _engine
    if _engine is None:
        _engine = create_async_engine(f"sqlite+aiosqlite:///{_db_path()}")
    return _engine


async def _ensure_schema() -> None:
    """Create the tables on first DB use (idempotent)."""
    global _schema_ready
    if _schema_ready:
        return
    async with _engine_for_db().begin() as conn:
        await conn.run_sync(SQLModel.metadata.create_all)
    _schema_ready = True


# --- Password encryption at rest (AES-256-GCM) ----------------------------
# The key comes from DB_ENCRYPTION_KEY (base64, 32 bytes) or a generated local
# key file next to the DB (gitignored). Stored values are base64(iv ‖ ciphertext),
# where AES-GCM appends its 16-byte tag to the ciphertext.


def _key_path() -> Path:
    env = os.environ.get("DB_KEY_PATH")
    return Path(env) if env else Path(f"{_db_path()}.key")


_key: bytes | None = None


def _load_or_create_key() -> bytes:
    env_key = os.environ.get("DB_ENCRYPTION_KEY")
    if env_key:
        return base64.b64decode(env_key)
    path = _key_path()
    try:
        return path.read_bytes()
    except FileNotFoundError:
        raw = os.urandom(32)
        path.write_bytes(raw)
        os.chmod(path, 0o600)
        return raw


def _key_bytes() -> bytes:
    global _key
    if _key is None:
        _key = _load_or_create_key()
    return _key


def _encrypt_password(plain: str) -> str:
    iv = os.urandom(12)
    ct = AESGCM(_key_bytes()).encrypt(iv, plain.encode("utf-8"), None)
    return base64.b64encode(iv + ct).decode("ascii")


def _decrypt_password(stored: str) -> str:
    combined = base64.b64decode(stored)
    iv, ct = combined[:12], combined[12:]
    return AESGCM(_key_bytes()).decrypt(iv, ct, None).decode("utf-8")


@dataclass
class StoredConnection:
    name: str
    type: str
    config: ChConfig
    database: str | None


async def _save_active_connection(name: str, c: ChConfig, conn_type: str = "clickhouse") -> None:
    password = _encrypt_password(c.password)
    now = _now_ms()
    await _ensure_schema()
    async with AsyncSession(_engine_for_db()) as s:
        row = (await s.exec(select(Connection).where(Connection.name == name))).first()
        if row is None:
            row = Connection(
                name=name,
                type=conn_type,
                host=c.host,
                port=c.port,
                username=c.username,
                password=password,
                last_active_at=now,
            )
        else:
            # Upsert by name; the selected database is intentionally left as-is.
            row.type = conn_type
            row.host = c.host
            row.port = c.port
            row.username = c.username
            row.password = password
            row.last_active_at = now
        s.add(row)
        await s.commit()


async def _save_selected_database(name: str, database: str) -> None:
    await _ensure_schema()
    async with AsyncSession(_engine_for_db()) as s:
        row = (await s.exec(select(Connection).where(Connection.name == name))).first()
        if row is not None:
            row.database = database
            s.add(row)
            await s.commit()


def _row_to_stored(row: Connection | None) -> StoredConnection | None:
    if row is None:
        return None
    try:
        password = _decrypt_password(row.password)
    except Exception:
        # Unreadable (key changed / legacy plaintext) — treat as unavailable.
        return None
    return StoredConnection(
        name=row.name,
        type=row.type,
        config=ChConfig(
            host=row.host, port=row.port, username=row.username, password=password
        ),
        database=row.database,
    )


async def _latest_active_connection() -> StoredConnection | None:
    await _ensure_schema()
    async with AsyncSession(_engine_for_db()) as s:
        row = (
            await s.exec(
                select(Connection).order_by(Connection.last_active_at.desc()).limit(1)
            )
        ).first()
        return _row_to_stored(row)


async def _connection_by_name(name: str) -> StoredConnection | None:
    await _ensure_schema()
    async with AsyncSession(_engine_for_db()) as s:
        row = (await s.exec(select(Connection).where(Connection.name == name))).first()
        return _row_to_stored(row)


async def _touch_connection(name: str) -> None:
    await _ensure_schema()
    async with AsyncSession(_engine_for_db()) as s:
        row = (await s.exec(select(Connection).where(Connection.name == name))).first()
        if row is not None:
            row.last_active_at = _now_ms()
            s.add(row)
            await s.commit()


def _now_ms() -> int:
    return int(time.time() * 1000)


# --- Sessions (one active connection per session, keyed by a cookie) ------


@dataclass
class _SessionState:
    name: str
    type: str
    config: ChConfig
    databases: list[str]
    database: str | None


# Cap the session map and evict the least-recently-used entry so it can't grow
# unbounded (every fresh cookie adds one). An evicted session transparently
# rebuilds on its next request via _ensure_session. OrderedDict keeps insertion
# order, so re-inserting on access moves an entry to the most-recently-used end.
_sessions: "OrderedDict[str, _SessionState]" = OrderedDict()
MAX_SESSIONS = int(os.environ.get("MAX_SESSIONS", "1000"))


def _get_session_entry(sid: str) -> _SessionState | None:
    s = _sessions.get(sid)
    if s is not None:
        _sessions.move_to_end(sid)
    return s


def _set_session_entry(sid: str, state: _SessionState) -> None:
    _sessions[sid] = state
    _sessions.move_to_end(sid)
    while len(_sessions) > MAX_SESSIONS:
        _sessions.popitem(last=False)


async def _build_session(
    name: str, config: ChConfig, database: str | None, conn_type: str = "clickhouse"
) -> tuple[_SessionState | None, str | None]:
    """List a connection's databases and build a session object."""
    ok, result = await list_databases(config)
    if not ok:
        return None, result  # type: ignore[return-value]
    databases: list[str] = result  # type: ignore[assignment]
    return (
        _SessionState(
            name=name,
            type=conn_type,
            config=config,
            databases=databases,
            database=database if database and database in databases else None,
        ),
        None,
    )


async def _ensure_session(sid: str) -> None:
    """At session start (a cookie we haven't seen), reconnect the latest active
    connection so a fresh session resumes where the last one left off."""
    if _get_session_entry(sid):
        return
    stored = await _latest_active_connection()
    if stored is None:
        return
    state, _ = await _build_session(stored.name, stored.config, stored.database, stored.type)
    if state is not None:
        _set_session_entry(sid, state)


async def get_session(sid: str) -> dict[str, Any]:
    """This session's state; auto-connects the latest active for an unseen cookie."""
    await _ensure_session(sid)
    s = _get_session_entry(sid)
    if s is None:
        return {"connected": False}
    return {
        "connected": True,
        "name": s.name,
        "type": s.type,
        "databases": s.databases,
        "database": s.database,
    }


async def connect_new(sid: str, name: str, config: ChConfig) -> dict[str, Any]:
    """Create: open a config, save + activate it for this session."""
    state, message = await _build_session(name, config, None, "clickhouse")
    if state is None:
        return {"ok": False, "message": message}
    _set_session_entry(sid, state)
    await _save_active_connection(name, config, "clickhouse")
    return {"ok": True, "name": name, "type": state.type, "databases": state.databases}


async def open_saved(sid: str, name: str) -> dict[str, Any]:
    """Open a saved connection by name for this session."""
    stored = await _connection_by_name(name)
    if stored is None:
        return {
            "ok": False,
            "message": f'no connection named "{name}"',
            "not_found": True,
        }
    # Reset the database so `connect <name>` always lands on the picker.
    state, message = await _build_session(stored.name, stored.config, None, stored.type)
    if state is None:
        return {"ok": False, "message": message}
    _set_session_entry(sid, state)
    await _touch_connection(name)
    return {"ok": True, "name": name, "type": state.type, "databases": state.databases}


async def select_database(sid: str, database: str) -> dict[str, Any]:
    """Select this session's active connection's database."""
    s = _get_session_entry(sid)
    if s is None:
        return {"ok": False, "message": "not connected", "reason": "no-session"}
    if not database or database not in s.databases:
        return {"ok": False, "message": "unknown database", "reason": "unknown"}
    s.database = database
    await _save_selected_database(s.name, database)
    return {"ok": True}


def _build_order_by(order_by: list[dict[str, Any]] | None) -> str:
    """Build an `ORDER BY` clause from `[{"name", "dir"}]`. Names are backtick-quoted
    (doubling any embedded backtick) and directions are whitelisted to ASC/DESC, so a
    malformed entry can't inject SQL. Empty/absent input yields no clause."""
    if not order_by:
        return ""
    parts: list[str] = []
    for col in order_by:
        if not isinstance(col, dict):
            continue
        name = col.get("name")
        if not isinstance(name, str) or not name:
            continue
        raw_dir = col.get("dir")
        direction = raw_dir.upper() if isinstance(raw_dir, str) else ""
        if direction not in ("ASC", "DESC"):
            direction = "ASC"
        escaped = name.replace("`", "``")
        parts.append(f"`{escaped}` {direction}")
    if not parts:
        return ""
    return "ORDER BY " + ", ".join(parts)


async def describe_query(sid: str, sql: str) -> dict[str, Any]:
    """Describe a query's output columns against this session's selected database."""
    await _ensure_session(sid)
    s = _get_session_entry(sid)
    if s is None:
        return {"ok": False, "message": "not connected", "reason": "no-session"}
    if not s.database:
        return {"ok": False, "message": "select a database first", "reason": "no-database"}
    ok, result = await ch_describe_query(s.config, sql, s.database)
    if not ok:
        return {"ok": False, "message": result}
    return {"ok": True, "fields": result}


# A dashboard query's result is capped at this many rows, matching the 1000-row
# ceiling on /api/clickhouse/query. Applied as the LIMIT of the paginating
# subselect that wraps each query.
DASHBOARD_ROW_CAP = 1000


def _parse_tsv_columns(text: str) -> dict[str, list[str]]:
    """Parse TabSeparatedWithNames into a column-oriented, insertion-ordered dict
    `{column_name: [values, …]}`. The first line is the column names; the rest
    are rows. Empty output yields an empty dict."""
    if text == "":
        return {}
    lines = text.split("\n")
    names = lines[0].split("\t")
    cols: dict[str, list[str]] = {name: [] for name in names}
    for line in lines[1:]:
        values = line.split("\t")
        for i, name in enumerate(names):
            cols[name].append(values[i] if i < len(values) else "")
    return cols


async def run_queries_for_connection(
    name: str, queries: dict[str, str]
) -> dict[str, Any]:
    """Run a dashboard's named queries against a saved connection by name,
    decoupled from any session/cookie. Fail-fast: an unknown connection, a
    connection with no selected database, or the first failing query aborts the
    whole call. On full success returns {"ok": True, "results": {name: {col:
    [values, …]}}} — column-oriented dicts ready for window.queries."""
    stored = await _connection_by_name(name)
    if stored is None:
        return {
            "ok": False,
            "reason": "no-connection",
            "message": f'no connection named "{name}"',
        }
    if not stored.database:
        return {
            "ok": False,
            "reason": "no-database",
            "message": (
                f'connection "{name}" has no selected database — select one for it '
                "or fully-qualify table names as db.table"
            ),
        }
    results: dict[str, dict[str, list[str]]] = {}
    for qname, sql in queries.items():
        inner = sql.rstrip().rstrip(";")
        paginated = f"SELECT * FROM (\n{inner}\n) LIMIT {DASHBOARD_ROW_CAP}"
        r = await ch_query(
            stored.config, paginated, database=stored.database, fmt="TabSeparatedWithNames"
        )
        if not r.ok:
            return {"ok": False, "reason": "query", "message": f"{qname}: {r.value}"}
        results[qname] = _parse_tsv_columns(r.value)
    return {"ok": True, "results": results}


async def run_query(
    sid: str,
    sql: str,
    limit: int,
    offset: int,
    fmt: str,
    order_by: list[dict[str, Any]] | None = None,
) -> dict[str, Any]:
    """Run a paginated SQL query against this session's selected database.
    Pagination wraps the query in a subselect so any SELECT can be paged; an optional
    `order_by` sorts that wrapper server-side."""
    await _ensure_session(sid)
    s = _get_session_entry(sid)
    if s is None:
        return {"ok": False, "message": "not connected", "reason": "no-session"}
    if not s.database:
        return {"ok": False, "message": "select a database first", "reason": "no-database"}
    inner = sql.rstrip().rstrip(";")
    clauses = [f"SELECT * FROM (\n{inner}\n)"]
    order_clause = _build_order_by(order_by)
    if order_clause:
        clauses.append(order_clause)
    clauses.append(f"LIMIT {int(limit)} OFFSET {int(offset)}")
    paginated = " ".join(clauses)
    r = await ch_query(s.config, paginated, database=s.database, fmt=fmt)
    if not r.ok:
        return {"ok": False, "message": r.value}
    return {"ok": True, "output": r.value}

File: backend/connect.ts
// Connection domain: the SQLite connection store (passwords encrypted at rest)
// and per-session active connections. No HTTP-server concerns live here; the
// operations return plain results and main.ts maps them to responses.

import { fromFileUrl } from "jsr:@std/path@^1.0.0"
import { decodeBase64, encodeBase64 } from "jsr:@std/encoding@^1/base64"
import { DatabaseSync } from "node:sqlite"
import { type ChConfig, listDatabases } from "./clickhouse.ts"

// --- Storage (SQLite) -----------------------------------------------------
const DB_PATH = Deno.env.get("DB_PATH") ??
  fromFileUrl(new URL("./queryview.db", import.meta.url))
const db = new DatabaseSync(DB_PATH)
db.exec(`
  CREATE TABLE IF NOT EXISTS connections (
    id             INTEGER PRIMARY KEY AUTOINCREMENT,
    name           TEXT NOT NULL UNIQUE,
    host           TEXT NOT NULL,
    port           INTEGER NOT NULL,
    username       TEXT NOT NULL,
    password       TEXT NOT NULL,
    database       TEXT,
    last_active_at INTEGER NOT NULL
  )
`)

// --- Password encryption at rest (AES-256-GCM) ----------------------------
// The key comes from DB_ENCRYPTION_KEY (base64, 32 bytes) or a generated local
// key file next to the DB (gitignored). Stored values are base64(iv ‖ ciphertext).
const KEY_PATH = Deno.env.get("DB_KEY_PATH") ?? `${DB_PATH}.key`

async function loadOrCreateKey(): Promise<CryptoKey> {
  let raw: Uint8Array
  const envKey = Deno.env.get("DB_ENCRYPTION_KEY")
  if (envKey) {
    raw = decodeBase64(envKey)
  } else {
    try {
      raw = await Deno.readFile(KEY_PATH)
    } catch {
      raw = crypto.getRandomValues(new Uint8Array(32))
      await Deno.writeFile(KEY_PATH, raw, { mode: 0o600 })
    }
  }
  return await crypto.subtle.importKey("raw", new Uint8Array(raw), "AES-GCM", false, [
    "encrypt",
    "decrypt",
  ])
}
const encryptionKey = await loadOrCreateKey()

async function encryptPassword(plain: string): Promise<string> {
  const iv = crypto.getRandomValues(new Uint8Array(12))
  const ct = new Uint8Array(
    await crypto.subtle.encrypt(
      { name: "AES-GCM", iv },
      encryptionKey,
      new Uint8Array(new TextEncoder().encode(plain)),
    ),
  )
  const combined = new Uint8Array(iv.length + ct.length)
  combined.set(iv, 0)
  combined.set(ct, iv.length)
  return encodeBase64(combined)
}

async function decryptPassword(stored: string): Promise<string> {
  const combined = new Uint8Array(decodeBase64(stored))
  const pt = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: combined.subarray(0, 12) },
    encryptionKey,
    combined.subarray(12),
  )
  return new TextDecoder().decode(pt)
}

type StoredConnection = ChConfig & { name: string; database: string | null }

async function saveActiveConnection(name: string, c: ChConfig): Promise<void> {
  const password = await encryptPassword(c.password)
  db.prepare(
    `INSERT INTO connections (name, host, port, username, password, last_active_at)
     VALUES (?, ?, ?, ?, ?, ?)
     ON CONFLICT(name) DO UPDATE SET
       host = excluded.host, port = excluded.port,
       username = excluded.username, password = excluded.password,
       last_active_at = excluded.last_active_at`,
  ).run(name, c.host, c.port, c.username, password, Date.now())
}

function saveSelectedDatabase(name: string, database: string): void {
  db.prepare(`UPDATE connections SET database = ? WHERE name = ?`)
    .run(database, name)
}

async function rowToConnection(
  row: Record<string, unknown> | undefined,
): Promise<StoredConnection | null> {
  if (!row) return null
  let password: string
  try {
    password = await decryptPassword(String(row.password))
  } catch {
    // Unreadable (key changed / legacy plaintext) — treat as unavailable.
    return null
  }
  return {
    name: String(row.name),
    host: String(row.host),
    port: Number(row.port),
    username: String(row.username),
    password,
    database: row.database == null ? null : String(row.database),
  }
}

function latestActiveConnection(): Promise<StoredConnection | null> {
  return rowToConnection(
    db.prepare(
      `SELECT name, host, port, username, password, database
       FROM connections ORDER BY last_active_at DESC LIMIT 1`,
    ).get() as Record<string, unknown> | undefined,
  )
}

function connectionByName(name: string): Promise<StoredConnection | null> {
  return rowToConnection(
    db.prepare(
      `SELECT name, host, port, username, password, database
       FROM connections WHERE name = ?`,
    ).get(name) as Record<string, unknown> | undefined,
  )
}

function touchConnection(name: string): void {
  db.prepare(`UPDATE connections SET last_active_at = ? WHERE name = ?`)
    .run(Date.now(), name)
}

// --- Sessions (one active connection per session, keyed by a cookie) ------
type Session = {
  name: string
  config: ChConfig
  databases: string[]
  database: string | null
}
const sessions = new Map<string, Session>()

// List a connection's databases and build a session object.
async function buildSession(
  name: string,
  config: ChConfig,
  database: string | null,
): Promise<{ ok: true; session: Session } | { ok: false; message: string }> {
  const r = await listDatabases(config)
  if (!r.ok) return { ok: false, message: r.message }
  const databases = r.databases
  return {
    ok: true,
    session: {
      name,
      config,
      databases,
      database: database && databases.includes(database) ? database : null,
    },
  }
}

// At session start (a cookie we haven't seen), reconnect the latest active
// connection so a fresh session resumes where the last one left off.
async function ensureSession(sid: string): Promise<void> {
  if (sessions.has(sid)) return
  const stored = await latestActiveConnection()
  if (!stored) return
  const { name, database, ...config } = stored
  const built = await buildSession(name, config, database)
  if (built.ok) sessions.set(sid, built.session)
}

export type SessionView =
  | { connected: true; name: string; databases: string[]; database: string | null }
  | { connected: false }

/** This session's state; auto-connects the latest active for an unseen cookie. */
export async function getSession(sid: string): Promise<SessionView> {
  await ensureSession(sid)
  const s = sessions.get(sid)
  return s
    ? { connected: true, name: s.name, databases: s.databases, database: s.database }
    : { connected: false }
}

type OpenResult =
  | { ok: true; name: string; databases: string[] }
  | { ok: false; message: string; notFound?: boolean }

/** Create: open a config, save + activate it for this session. */
export async function connectNew(
  sid: string,
  name: string,
  config: ChConfig,
): Promise<OpenResult> {
  const built = await buildSession(name, config, null)
  if (!built.ok) return { ok: false, message: built.message }
  sessions.set(sid, built.session)
  await saveActiveConnection(name, config)
  return { ok: true, name, databases: built.session.databases }
}

/** Open a saved connection by name for this session. */
export async function openSaved(sid: string, name: string): Promise<OpenResult> {
  const stored = await connectionByName(name)
  if (!stored) {
    return { ok: false, message: `no connection named "${name}"`, notFound: true }
  }
  const { name: _n, database: _d, ...config } = stored
  // Reset the database so `connect <name>` always lands on the picker.
  const built = await buildSession(name, config, null)
  if (!built.ok) return { ok: false, message: built.message }
  sessions.set(sid, built.session)
  touchConnection(name)
  return { ok: true, name, databases: built.session.databases }
}

/** Select this session's active connection's database. */
export function selectDatabase(
  sid: string,
  database: string,
): { ok: true } | { ok: false; message: string; reason: "no-session" | "unknown" } {
  const s = sessions.get(sid)
  if (!s) return { ok: false, message: "not connected", reason: "no-session" }
  if (!database || !s.databases.includes(database)) {
    return { ok: false, message: "unknown database", reason: "unknown" }
  }
  s.database = database
  saveSelectedDatabase(s.name, database)
  return { ok: true }
}
